### ISSUE TITLE
Editing line on stocktake wipes batch counts

### ISSUE THREAD

**Issue by Chris-Petty** [openmsupply/mobile#1471]
## Describe the bug

When editing a stocktake, you can go item by item and/or edit batches of an item. When editing by item, it does some dark magic system related to the fabled "FEFO" prophecy.

3.0.0 rewrite included saving changes to a cell on every keystroke. If you delete a number before entering a new one, it'll apply changing stock to 0 across all batches. Then as you type in a new number, it only adds to the first expiring (I think). This is bad, as it ruins all your batch stock counts and movements if you then finalised the stocktake.

### To reproduce

  1. Go to a Stocktake that has some lines with existing counts
  2. Inspect batch quantities for a line with stock. Add some extra batches if there is only 1, and add some stock to them
  3. Go back to the item list and edit the item stock, say from 100 to 103:
    - Tap the cell (which focuses the number) and start typing a new one. This will wipe all stock then add the count to the first expiring batch (or last? I'm tired)
    - Tap the cell and set the cursor to the end. backspace once to __10__, removing 90 stock from batches according to reverse FEFO (??? (the order doesn't actually matter regarding this bug)), then adding 93 to the first expiring batch*
  4. Check batches again and see all the stock has shifted violently to the nearest expiry.

_*This method will work even on 100->10->100. Probably not expecting much to change, but it do_

### Expected behaviour

Batch quantities are preserved if you are trying to edit an item counted quantity. 

Basically I think cells in stocktake items should only callback on losing focus rather than on every keystroke. Not sure how else you'd fix this. Also whoops, didn't see this one coming.

### Version and device info

- App version: 3.0.1
- Tablet model: TabA SM-T580
- OS version: YEETDROID

### Additional context

#625 can probably be closed

**Comment by andreievg** [comment]
Hi @Chris-Petty, this would only occur when `new` batches are added (initial stocktake most likely). If there are multiple batches with snapshot quantity, changing total quantity should apply batches quantity correctly.

Making stateful component with on focus loss update would also work, but i don't see any way to prevent value loss if finalised or back button is clicked when number is entered in the stateful component (which i think might be worse).

It's a bit hard to optimise total quantity to alway applying correctly to correct batches, and we should emphasise this to the users (i.e. either use by batch or total quantity, please don't mix).

Can we please park this until you get back for more thorough discussion ?

Trick question: I have 5 green apples and 10 red apples, so that's 15 apples. oops actually i have 12 apples, how many green and red apples do i have ?

**Comment by Chris-Petty** [comment]
## Foreword
@andreievg is right. Snapshot quantity does protect your old stock. You carefully edited new batches you added, then edit the overall, you'll still lose that work. Still not so bad cause of snapshots though (my late night training Mark didn't have snapshots > 0, so didn't pick it up!)

I think it'd still be an improvement if it action to apply changes could be separated from reducer data state or perhaps debounced a few seconds so this doesn't happen, wouldn't need to be stateful. Doesn't need to be immediate though

## A story in pictures:
<img width="1008" alt="Screen Shot 2019-10-31 at 13 36 02" src="https://user-images.githubusercontent.com/7684221/67917111-b4bddb00-fbe3-11e9-9a4a-0de1b348d63a.png">
<img width="959" alt="Screen Shot 2019-10-31 at 13 36 19" src="https://user-images.githubusercontent.com/7684221/67917113-b7203500-fbe3-11e9-82be-6d36e8ea61b4.png">
<img width="436" alt="Screen Shot 2019-10-31 at 13 36 32" src="https://user-images.githubusercontent.com/7684221/67917118-bc7d7f80-fbe3-11e9-9ab9-3f772da765d4.png">
<img width="388" alt="Screen Shot 2019-10-31 at 13 36 40" src="https://user-images.githubusercontent.com/7684221/67917127-c0a99d00-fbe3-11e9-8d9e-6b54bdaf4dbf.png">
<img width="945" alt="Screen Shot 2019-10-31 at 13 36 51" src="https://user-images.githubusercontent.com/7684221/67917134-c606e780-fbe3-11e9-95fc-974fe8b86a16.png">

**Comment by joshxg** [comment]
I think you're both right and both situations are undesired. 

"Both" being: Save on keystroke (Causing this issue) V Save on unfocus (Causing the problems andrei points out). 

I think Save on unfocus/debouncing or any other fix (hack) introduces inconsistency in behaviour - but I'm not sure what is worse: inconsistency or this quantity problem -  it's more of a everyone a small amount V a smaller amount of people a larger amount issue?

I imagine there's a solution to both problems, but in a form different to when or how a callback is invoked

**Comment by joshxg** [comment]
Closing the PR for this issue until implementation/solution is agreed upon
